man/bestups.8
.TH BESTUPS 8 "Wed Jul 28 2004" "" "Network UPS Tools (NUT)" 
.SH NAME  
bestups \- Driver for Best Power / SOLA (Phoenixtec protocol) UPS equipment
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
bestups driver.  For information about the core driver, see  
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
bestups was designed to monitor Best Power UPS hardware like the Fortress,
Fortress Telecom, Axxium Rackmount and Patriot Pro.  It also recognizes
and supports SOLA units such as the 325, 520 and 620.  In addition, the 
Best 610 is supported using the `ID' option.

Other UPS hardware using the Phoenixtec protocol should also work, but
they will generate a warning since their battery information is not known.

This driver does not support some older Best/SOLA units.

.SH EXTRA ARGUMENTS

This driver supports the following optional settings in the
\fBups.conf\fR(5):

.IP "nombattvolt=\fInum\fR"

Override the battery float voltage which is normally determined by
asking the hardware.  This is useful if your UPS constantly reports
battery.charge values just below 100% even when it's completely charged.

If you have this problem, set this to whatever battery.voltage reports
when the UPS is known to be completely charged with a good battery.

The author's Best Fortress 750 uses nombattvolt=27.4.

.IP "battvoltmult=\fInum\fR"

Multiply the reported battery voltage by this number. Some devices
report only a fraction of the total battery voltage.

For example, the SOLA 610 700VA UPS (with a 24V battery) reports the
single cell voltage (about 2.27V when fully charged). In this particular
case you can set 'battvoltmult = 12' in \fBups.conf\fR(8) to fix this.

.IP "ID=\fIstring\fR"

Set the Identification response string.  This should only be used
with hardware that supports the Phoenixtec protocol status inquiry 
commands, but not the "ID" command, such as the Best/SOLA 610.  Format 
of the ID string is: AAA,BBBB,CCC,DDD,EE.E,FF.F

AAA is the three-character identification for the UPS model.

BBBB is the output power in VA (volt amperes). B is an integer number 
ranging from 0 to 9.

CCC is the Nominal Input Voltage. C is an integer number ranging from 0 
to 9. The unit is Volts AC.

DDD is the Nominal Output Voltage. D is an integer number ranging from 0 
to 9. The unit is Volts AC.

EE.E is the Battery Voltage that will cause the UPS to shut itself off. 
E is an integer number ranging from 0 to 9. Then unit is Volts DC and a 
decimal point is present.

FF.F or FFF.F is the Battery Voltage at full charge. F is an integer 
number ranging from 0 to 9. Then unit is Volts DC. Typically, for 700VA, 
1KVA and 1.5KVA units, the format is FF.F. For 2KVA and 3KVA units, the 
format is FFF.F.

Example: a Best 610 1.5KVA unit would use the string 
"610,1500,120,120,10.0,48.0".

.SH BUGS

The battery charge percentage value (in battery.charge) is derived from
the voltage data that the UPS returns, since the UPS doesn't return that
value directly.  On some hardware, the charge will remain at 100% for a
long time and then drops quickly shortly before the battery runs out.
You can confirm from the battery.voltage readings that this is a problem
with the UPS and not this driver. 

Similarly, the float from the charger in some models forces the battery
charge percentage back up to 100% immedately after the UPS goes back
on\(hyline, so you can't tell when it is really recharged.

Finally, some models give one value for the battery's nominal voltage and
yet actually have a nominal voltage slightly below that.  This leads to
things such as the perpetual 98.7% charge on the author's Fortress 750,
even when it's been charging for weeks.  You can use nombattvolt= in 
\fBups.conf\fR(8) to fix this.

.SH AUTHOR
Russell Kroll, Jason White

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.

 - the megatec and megatec_usb drivers have also been backported from the
   development (trunk) version.
   
 - the client development files have also received some care:
   the upsclient pkg-config file has been fixed, and the upsclient.h
   file allows older NUT clients to continue using the UPSCONN structure.
   
 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.0 - what's new since 2.0.5:

 - The new build infrastructure, using automake, is now used.
   This has major impact on the compilation and installation procedures,
   and thus on the NUT packaging.
   For more information, refer to UPGRADING and packaging/debian/ for
   an example of migration.

 - NUT now provides support for FreeDesktop Hardware Abstraction Layer
   (HAL) which brings full Plug And Play experience to USB UPS owners.
   For more information, refer to docs/nut-hal.txt.

 - support for new devices: Ablerex 625L, ActivePower 400VA, 2000VA;
   Belkin Home Office F6H350-SER, F6H500-SER, F6H650-SER; Belkin Office
   Series F6C550-AVR; Belkin Universal UPS F6C100-UNV (USB), F6C1100-UNV
   (USB), F6C1200-UNV (USB), F6H350deUNV (serial), F6H350ukUNV (serial),
   F6H650ukUNV (serial); Compaq R3000h; Cyber Power Systems PR2200;
   Dynex DX-800U; Geek Squad GS1285U; Krauler UP-M500VA; Mecer ME-2000;
   MGE UPS SYSTEMS Ellipse MAX; Online Zinto D; PowerTech SMK-800; SVEN
   Power Pro+ series, Power Smart RM 2000; Tripp-Lite SmartOnline
   SU1500RTXL2ua, smart2200RMXL2U.

 - added IPv6 support,

 - the newmge-shut driver has appeared. This one uses the same HID core
   as usbhid-ups, but communicate over a serial link. It will eventually
   replace the current mge-shut driver.

 - client commands (upsc, upsrw and upscmd): hostname is now optional,
   and defaults to "localhost"

 - many drivers have been improved and have received bug fixes:
   powerpanel, megatec, megatec_usb, safenet, tripplite_usb, gamatronic,

 - the hotplug and udev scripts, in charge of setting the right
   permissions on the USB devices, are now installed automatically
   when appropriate.

 - more generally, the NUT core and documentation, including the manpages,
   have been improved and updated.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550; Eltek; Inform
   GUARD; Microsol Rhino; Opti-UPS PowerES 420E; PowerMan RealSmart,
   BackPro; Powerware PW9315 3-phase; SOLA 305; Tripp-Lite
   SMART550USB, SMART2200RMXL2U, OMNI1000LCD, OMNI900LCD, OMNI650LCD,
   1500 LCD, AVR550U; Viewsonic PowerES 420E.

 - bcmxcp: added 3-phase support

 - megatec: better hardware support, more instant commands

 - mge-hid: support more instant commands

 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices

 - solis: shutdown programming, support new cables, Solaris support

 - tripplite_usb: updated SMARTPRO support, fixed OL/OB reporting, 
   better error handling, some memory bugs

 - new dummy-ups driver simulator

 - added HTML interface for access to CGI scripts

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.4 - what's new since 2.0.3:

 - The newhidups critical bug (segmentation fault) has been fixed. It has
   also received some more care, like buxfixes and new models support and
   enhancement for Solaris.
   [Peter Selinger and Arnaud Quette]

 - A bug has been fixed in NUT core to support resuming from suspend-to-disk.
   This should also fix other similar issues, like time synchronisation
   through the NTP - Network Time Protocol.
   [Arjen de Korte]

 - The mge-shut driver now better detects the Low Battery status, support
   new models and fixes some wrong status and data. It also fixes some
   issue where the UPS wasn't restarting (refer to mge-shut manpage).
   [Arnaud Quette]

 - The genericups custom configuration through ups.conf is working again
   [Arjen de Korte]

 - The genericups driver type 22 also support CyberPower 725SL
   (and maybe others SL models)
   [David Kaufman]

 - The new megatec driver, which will replace a bunch of drivers by nut 2.2
   (refer to docs/megatec.txt and UPGRADING) has been backported from the
   trunk (Development tree). The powermust driver has also received some
   attention.
   [Carlos Rodrigues]

 - The new rhino driver was added to support Microsol Rhino UPS hardware

drivers/bestups.c
/* bestups.c - model specific routines for Best-UPS Fortress models

   Copyright (C) 1999  Russell Kroll <rkroll@exploits.org>

   ID config option by Jason White <jdwhite@jdwhite.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include "main.h"
#include "serial.h"
#include "bestups.h"

#define ENDCHAR  13	/* replies end with CR */
#define MAXTRIES 5
#define UPSDELAY 50000	/* 50 ms delay required for reliable operation */

#define SER_WAIT_SEC	3	/* allow 3.0 sec for ser_get calls */
#define SER_WAIT_USEC	0

static	float	lowvolt = 0, highvolt = 0;
static	int	battvoltmult = 1;
static	int	inverted_bypass_bit = 0;

static void model_set(const char *abbr, const char *rating)
{
	if (!strcmp(abbr, "FOR")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "Fortress %s", rating);
		return;
	}

	if (!strcmp(abbr, "FTC")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "Fortress Telecom %s", rating);
		return;
	}

	if (!strcmp(abbr, "PRO")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "Patriot Pro %s", rating);
		inverted_bypass_bit = 1;
		return;
	}

	if (!strcmp(abbr, "PR2")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "Patriot Pro II %s", rating);
		inverted_bypass_bit = 1;
		return;
	}

	if (!strcmp(abbr, "325")) {
		dstate_setinfo("ups.mfr", "%s", "Sola Australia");
		dstate_setinfo("ups.model", "Sola 325 %s", rating);
		return;
	}

	if (!strcmp(abbr, "520")) {
		dstate_setinfo("ups.mfr", "%s", "Sola Australia");
		dstate_setinfo("ups.model", "Sola 520 %s", rating);
		return;
	}

	if (!strcmp(abbr, "610")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "610 %s", rating);
		return;
	}

	if (!strcmp(abbr, "620")) {
		dstate_setinfo("ups.mfr", "%s", "Sola Australia");
		dstate_setinfo("ups.model", "Sola 620 %s", rating);
		return;
	}

	if (!strcmp(abbr, "AX1")) {
		dstate_setinfo("ups.mfr", "%s", "Best Power");
		dstate_setinfo("ups.model", "Axxium Rackmount %s", rating);
		return;
	}

	dstate_setinfo("ups.mfr", "%s", "Unknown");
	dstate_setinfo("ups.model", "Unknown %s (%s)", abbr, rating);

	printf("Unknown model detected - please report this ID: '%s'\n", abbr);
}

static int instcmd(const char *cmdname, const char *extra)
{
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send_pace(upsfd, UPSDELAY, "CT\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "test.battery.start")) {
		ser_send_pace(upsfd, UPSDELAY, "T\r");
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}

static int get_ident(char *buf, size_t bufsize)
{
	int	i, ret;
	char	*ID;

	ID = getval("ID");	/* user-supplied override from ups.conf */

	if (ID) {
		upsdebugx(2, "NOTE: using user-supplied ID response");
		snprintf(buf, bufsize, "%s", ID);
		return 1;
	}

	for (i = 0; i < MAXTRIES; i++) {
		ser_send_pace(upsfd, UPSDELAY, "\rID\r");

		ret = ser_get_line(upsfd, buf, bufsize, ENDCHAR, "", 
			SER_WAIT_SEC, SER_WAIT_USEC);

		if (ret > 0)
			upsdebugx(2, "get_ident: got [%s]", buf);

		/* buf must start with ( and be in the range [25-27] */
		if ((ret > 0) && (buf[0] != '(') && (strlen(buf) >= 25) &&
			(strlen(buf) <= 27))
			return 1;

		sleep(1);
	}

	upslogx(LOG_INFO, "Giving up on hardware detection after %d tries",
		MAXTRIES);

	return 0;
}

static void ups_ident(void)
{
	int	i;
	char	buf[256], *ptr;
	char	*model = NULL, *rating = NULL;

	if (!get_ident(buf, sizeof(buf))) {
		fatalx(EXIT_FAILURE, "Unable to detect a Best/SOLA or Phoenix protocol UPS");
	}

	/* FOR,750,120,120,20.0,27.6 */
	ptr = strtok(buf, ",");

	for (i = 0; ptr; i++) {

		switch (i)
		{
		case 0:
			model = ptr;
			break;

		case 1:
			rating = ptr;
			break;

		case 2:
			dstate_setinfo("input.voltage.nominal", "%d", atoi(ptr));
			break;

		case 3:
			dstate_setinfo("output.voltage.nominal", "%d", atoi(ptr));
			break;

		case 4:
			lowvolt = atof(ptr);
			break;

		case 5:
			highvolt = atof(ptr);
			break;
		}

		ptr = strtok(NULL, ",");
	}

	if ((!model) || (!rating)) {
		fatalx(EXIT_FAILURE, "Didn't get a valid ident string");
	}
	
	model_set(model, rating);

	/* Battery voltage multiplier */
	ptr = getval("battvoltmult");

	if (ptr) {
		battvoltmult = atoi(ptr);
	}

	/* Lookup the nominal battery voltage (should be between lowvolt and highvolt */
	for (i = 0; i < 8; i++) {
		const int	nominal[] = { 2, 6, 12, 24, 36, 48, 72, 96 };

		if ((lowvolt < nominal[i]) && (highvolt > nominal[i])) {
			dstate_setinfo("battery.voltage.nominal", "%d", battvoltmult * nominal[i]);
			break;
	 	}
	}

	ptr = getval("nombattvolt");

	if (ptr) {
		highvolt = atof(ptr);
	}
}

static void ups_sync(void)
{
	char	buf[256];
	int	i, ret;

	for (i = 0; i < MAXTRIES; i++) {
		ser_send_pace(upsfd, UPSDELAY, "\rQ1\r");

		ret = ser_get_line(upsfd, buf, sizeof(buf), ENDCHAR, "", 
			SER_WAIT_SEC, SER_WAIT_USEC);

		/* return once we get something that looks usable */
		if ((ret > 0) && (buf[0] == '('))
			return;

		usleep(250000);
	}

	fatalx(EXIT_FAILURE, "Unable to detect a Best/SOLA or Phoenix protocol UPS");
}

void upsdrv_initinfo(void)
{
	ups_sync();
	ups_ident();

	printf("Detected %s %s on %s\n", dstate_getinfo("ups.mfr"), 
		dstate_getinfo("ups.model"), device_path);

	/* paranoia - cancel any shutdown that might already be running */
	ser_send_pace(upsfd, UPSDELAY, "C\r");

	upsh.instcmd = instcmd;

	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);
	dstate_addcmd("test.battery.start");
	dstate_addcmd("test.battery.stop");
}

static int ups_on_line(void)
{
	int	i, ret;
	char	temp[256], pstat[32];

	for (i = 0; i < MAXTRIES; i++) {
		ser_send_pace(upsfd, UPSDELAY, "\rQ1\r");

		ret = ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 
			SER_WAIT_SEC, SER_WAIT_USEC);

		/* Q1 must return 46 bytes starting with a ( */
		if ((ret > 0) && (temp[0] == '(') && (strlen(temp) == 46)) {

			sscanf(temp, "%*s %*s %*s %*s %*s %*s %*s %s", pstat);

			if (pstat[0] == '0')
				return 1;	/* on line */

			return 0;	/* on battery */
		}

		sleep(1);
	}

	upslogx(LOG_ERR, "Status read failed: assuming on battery");

	return 0;	/* on battery */
}	

void upsdrv_shutdown(void)
{
	printf("The UPS will shut down in approximately one minute.\n");

	if (ups_on_line())
		printf("The UPS will restart in about one minute.\n");
	else
		printf("The UPS will restart when power returns.\n");

	ser_send_pace(upsfd, UPSDELAY, "S01R0001\r");
}

void upsdrv_updateinfo(void)
{
	char	involt[16], outvolt[16], loadpct[16], acfreq[16], 
		battvolt[16], upstemp[16], pstat[16], buf[256];
	float	bvoltp;
	int	ret;

	ret = ser_send_pace(upsfd, UPSDELAY, "\rQ1\r");

	if (ret < 1) {
		ser_comm_fail("ser_send_pace failed");
		dstate_datastale();
		return;
	}

	/* these things need a long time to respond completely */
	usleep(200000);

	ret = ser_get_line(upsfd, buf, sizeof(buf), ENDCHAR, "", 
		SER_WAIT_SEC, SER_WAIT_USEC);

	if (ret < 1) {
		ser_comm_fail(NULL);
		dstate_datastale();
		return;
	}

	if (ret < 46) {
		ser_comm_fail("Poll failed: short read (got %d bytes)", ret);
		dstate_datastale();
		return;
	}

	if (ret > 46) {
		ser_comm_fail("Poll failed: response too long (got %d bytes)",
			ret);
		dstate_datastale();
		return;
	}

	if (buf[0] != '(') {
		ser_comm_fail("Poll failed: invalid start character (got %02x)",
			buf[0]);
		dstate_datastale();
		return;
	}

	ser_comm_good();

	sscanf(buf, "%*c%s %*s %s %s %s %s %s %s", involt, outvolt, 
		loadpct, acfreq, battvolt, upstemp, pstat);

	/* Guesstimation of battery charge left (inaccurate) */
	bvoltp = 100 * (atof(battvolt) - lowvolt) / (highvolt - lowvolt);

	if (bvoltp > 100) {
		bvoltp = 100;
	}

	dstate_setinfo("battery.voltage", "%.1f", battvoltmult * atof(battvolt));
	dstate_setinfo("input.voltage", "%s", involt);
	dstate_setinfo("output.voltage", "%s", outvolt);
	dstate_setinfo("ups.load", "%s", loadpct);
	dstate_setinfo("input.frequency", "%s", acfreq);

	if(upstemp[0] != 'X') {
		dstate_setinfo("ups.temperature", "%s", upstemp);
	}

	dstate_setinfo("battery.charge", "%02.1f", bvoltp);

	status_init();

	if (pstat[0] == '0') {
		status_set("OL");		/* on line */

		/* only allow these when OL since they're bogus when OB */

		if (pstat[2] == (inverted_bypass_bit ? '0' : '1')) {
			/* boost or trim in effect */
			if (atof(involt) < atof(outvolt))
				status_set("BOOST");

			if (atof(involt) > atof(outvolt))
				status_set("TRIM");
		}

	} else {
		status_set("OB");		/* on battery */
	}

	if (pstat[1] == '1')
		status_set("LB");		/* low battery */

	status_commit();
	dstate_dataok();
}

void upsdrv_help(void)
{
}

void upsdrv_makevartable(void)
{
	addvar(VAR_VALUE, "nombattvolt", "Override nominal battery voltage");
	addvar(VAR_VALUE, "ID", "Force UPS ID response string");
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - Best UPS driver %s (%s)\n", 
		DRV_VERSION, UPS_VERSION);
}

void upsdrv_initups(void)
{
	upsfd = ser_open(device_path);
	ser_set_speed(upsfd, device_path, B2400);
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}
